File: cli/src/huerise_cli/commands/rooms.py
import typer

from huerise_cli.client import api_client
from huerise_cli.config import Config
from huerise_cli.generated.api.rooms_scenes import (
    activate_scene as activate_scene_api,
)
from huerise_cli.generated.api.rooms_scenes import (
    get_room as get_room_api,
)
from huerise_cli.generated.api.rooms_scenes import (
    list_rooms as list_rooms_api,
)
from huerise_cli.generated.models import RoomRead
from huerise_cli.output import (
    JsonOption,
    console,
    emit_json,
    emit_record,
    emit_table,
    unwrap,
)

app = typer.Typer(no_args_is_help=True, help="Browse rooms and Hue scenes.")


def _room_row(room: RoomRead) -> dict[str, object]:
    return {"name": room.name, "scenes": ", ".join(room.scene_names)}


@app.command("list")
def list_rooms(json_: JsonOption = False) -> None:
    """List every room Hue knows about."""
    with api_client(Config.from_env()) as client:
        rooms = unwrap(list_rooms_api.sync_detailed(client=client))

    if json_:
        emit_json([room.to_dict() for room in rooms])
    else:
        emit_table([_room_row(room) for room in rooms], empty_message="No rooms found.")


@app.command("get")
def get_room(room_name: str, json_: JsonOption = False) -> None:
    """Show a room and the scenes it offers."""
    with api_client(Config.from_env()) as client:
        room = unwrap(get_room_api.sync_detailed(client=client, room_name=room_name))

    if json_:
        emit_json(room.to_dict())
    else:
        emit_record(_room_row(room))


@app.command("activate-scene")
def activate_scene(room_name: str, scene_name: str) -> None:
    """Preview a scene the way an alarm would start it."""
    with api_client(Config.from_env()) as client:
        unwrap(
            activate_scene_api.sync_detailed(
                client=client, room_name=room_name, scene_name=scene_name
            )
        )
    console.print(f"[green]Activated '{scene_name}' in {room_name}.[/green]")

File: internal/cli/rooms.go
package cli

import (
	"fmt"
	"strings"

	"github.com/mathisarends/huerise/internal/api"
)

type roomsCommand struct {
	List          roomsListCommand          `cmd:"" help:"List every room Hue knows about."`
	Get           roomsGetCommand           `cmd:"" help:"Show a room and its available scenes."`
	ActivateScene roomsActivateSceneCommand `cmd:"" name:"activate-scene" help:"Preview a scene the way an alarm would start it."`
}

type roomsListCommand struct{}

type roomsGetCommand struct {
	RoomName string `arg:"" name:"room" help:"Hue room name."`
}

type roomsActivateSceneCommand struct {
	RoomName  string `arg:"" name:"room" help:"Hue room name."`
	SceneName string `arg:"" name:"scene" help:"Hue scene name."`
}

func (roomsListCommand) Run(runtime *Runtime) error {
	client, err := runtime.client()
	if err != nil {
		return err
	}
	rooms, err := client.ListRooms(runtime.ctx)
	if err != nil {
		return err
	}
	return runtime.output(rooms, func() error {
		rows := make([][]string, 0, len(rooms))
		for _, room := range rooms {
			rows = append(rows, []string{room.Name, strings.Join(room.SceneNames, ", ")})
		}
		return writeTable(runtime.stdout, []string{"NAME", "SCENES"}, rows, "No rooms found.")
	})
}

func (command roomsGetCommand) Run(runtime *Runtime) error {
	client, err := runtime.client()
	if err != nil {
		return err
	}
	response, err := client.GetRoom(runtime.ctx, api.GetRoomParams{RoomName: command.RoomName})
	if err != nil {
		return err
	}
	switch result := response.(type) {
	case *api.RoomRead:
		return runtime.output(result, func() error {
			return writeRecord(runtime.stdout,
				recordField{Name: "name", Value: result.Name},
				recordField{Name: "scenes", Value: strings.Join(result.SceneNames, ", ")},
			)
		})
	case *api.HTTPValidationError:
		return validationError(result)
	default:
		return fmt.Errorf("unexpected get room response %T", response)
	}
}

func (command roomsActivateSceneCommand) Run(runtime *Runtime) error {
	client, err := runtime.client()
	if err != nil {
		return err
	}
	response, err := client.ActivateScene(runtime.ctx, api.ActivateSceneParams{
		RoomName: command.RoomName, SceneName: command.SceneName,
	})
	if err != nil {
		return err
	}
	switch result := response.(type) {
	case *api.ActivateSceneNoContent:
		value := map[string]any{"activated": true, "room_name": command.RoomName, "scene_name": command.SceneName}
		return runtime.output(value, func() error {
			_, err := fmt.Fprintf(runtime.stdout, "Activated %q in %s.\n", command.SceneName, command.RoomName)
			return err
		})
	case *api.HTTPValidationError:
		return validationError(result)
	default:
		return fmt.Errorf("unexpected activate scene response %T", response)
	}
}
